http-server-netty/src/main/java/io/micronaut/http/server/netty/MicronautHttpData.java

```diff
@@ -0,0 +1,781 @@
+ * {@link #get()} and friends. However, if the user wants a reactive stream of data as it comes in,
+ * this class can release chunks of that data for concurrent access by the user (see
+ * {@link #pollChunk()}).<br>
+ * This class moves data to disk dynamically once the configured threshold is reached.
+ *
+ * @param <D> This {@link HttpData} type, for {@code return (D) this} on various methods
+ */
+@Internal
+public abstract sealed class MicronautHttpData<D extends HttpData> extends AbstractReferenceCounted implements HttpData {
+    @SuppressWarnings("rawtypes")
+    private static final Supplier<ResourceLeakDetector<MicronautHttpData>> LEAK_DETECTOR = SupplierUtil.memoized(() ->
+        ResourceLeakDetectorFactory.instance().newResourceLeakDetector(MicronautHttpData.class));
+
+    private static final Logger LOG = LoggerFactory.getLogger(MicronautHttpData.class);
+
+    private static final int MMAP_SEGMENT_SIZE = 1024 * 1024 * 1024;
+    private static final int MAX_CHUNK_SIZE = 1024 * 1024 * 1024;
+
+    final Factory factory;
+
+    long definedSize = 0;
+    Charset charset;
+
+    /**
+     * Additional data for {@link FormRouteCompleter}.
+     */
+    FormRouteCompleter.HttpDataAttachment attachment;
+
+    @Nullable
+    @SuppressWarnings("rawtypes")
+    private final ResourceLeakTracker<MicronautHttpData> tracker = LEAK_DETECTOR.get().track(this);
+
+    private final String name;
+
+    private final List<Chunk> chunks = new ArrayList<>();
+
+    private long size = 0;
+
+    @Nullable
+    private Path path;
+    private FileChannel channel;
+    private List<ByteBuf> mmapSegments;
+
+    private boolean completed = false;
+
+    private int pollIndex = 0;
+
+    private MicronautHttpData(Factory factory, String name) {
+        this.factory = factory;
+        this.name = name;
+        this.charset = factory.characterEncoding;
+        chunks.add(new Chunk(0));
+    }
+
+    private boolean shouldMoveToDisk(long newSize) {
+        if (factory.multipartConfiguration.isDisk()) {
+            return true;
+        } else if (factory.multipartConfiguration.isMixed()) {
+            return newSize >= factory.multipartConfiguration.getThreshold();
+        } else {
+            return false;
+        }
+    }
+
+    private Chunk lastChunk() {
+        return chunks.get(chunks.size() - 1);
+    }
+
+    /**
+     * Get a chunk of data. The chunk will have a fixed content, it will not be amended with
+     * further input.
+     *
+     * @return The chunk, or {@code null} if this data is {@link #isCompleted() completed} and all
+     * chunks have been polled.
+     */
+    public Chunk pollChunk() {
+        if (pollIndex >= chunks.size()) {
+            return null;
+        }
+        Chunk chunk = chunks.get(pollIndex++);
+        if (pollIndex == chunks.size() && !completed) {
+            chunks.add(new Chunk(size));
+        }
+        chunk.retain();
```

Comment:
where is this released?

Reply:
in Chunk.claim usually, i will add a comment